File: 2064.minimized-maximum-of-products-distributed-to-any-store.cpp
/*
 * @lc app=leetcode id=2064 lang=cpp
 *
 * [2064] Minimized Maximum of Products Distributed to Any Store
 */

// @lc code=start
#include <vector>
#include <algorithm>

class Solution {
public:
    int ceilDiv(int a, int b) {
        return (a + b - 1) / b;
    }

    int minimizedMaximum(int n, std::vector<int>& quantities) {
        int l = 1, r = *std::max_element(quantities.begin(), quantities.end());
        while (l < r) {
            int mid = l + (r - l) / 2;
            int numPeople = 0;
            for (int q : quantities) {
                numPeople += ceilDiv(q, mid);
            }
            if (numPeople <= n) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return l;
    }
};

// @lc code=end

File: 2064.minimized-maximum-of-products-distributed-to-any-store.py
#
# @lc app=leetcode id=2064 lang=python3
#
# [2064] Minimized Maximum of Products Distributed to Any Store
#

# @lc code=start
from typing import List


class Solution:
    '''
    Solve by binary search. The answer must be between 0 and max(quantities)
    To give at most x, quantities[i] has to be given to ceildiv(quantities[i], x) people.
    '''
    def ceildiv(self, a, b):
        return -(a // -b)

    def minimizedMaximum(self, n: int, quantities: List[int]) -> int:
        l, r = 1, max(quantities)
        while l < r:
            mid =  l + (r - l) // 2
            numPeople = 0
            for q in quantities: numPeople += self.ceildiv(q, mid)
            if numPeople <= n:
                r = mid
            else:
                l = mid + 1
        return l


# @lc code=end

